api/core/database/models/api/bot/user.py

```diff
@@ -0,0 +1,61 @@
+from typing import NoReturn, Union
+
+from sqlalchemy import ARRAY, BigInteger, Boolean, CheckConstraint, Column, SmallInteger, String
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker, validates
+
+from api.core.database import Base
+from api.core.settings import settings
+from .role import Role
+
+engine = create_engine(settings.database_url)
+SessionLocal = sessionmaker(bind=engine)
+
+
+class User(Base):
+    """A Discord user."""
+
+    __tablename__ = 'user'
+    __table_args__ = (
+        CheckConstraint('discriminator >= 0'),
+    )
+
+    # The ID of this user, taken from Discord.
+    id = Column(BigInteger, primary_key=True)
+
+    # The username, taken from Discord.
+    name = Column(String(32), nullable=False)
+
+    # Discriminators may not exceed `9999`."
+    discriminator = Column(SmallInteger, nullable=False)
+
+    # Whether this user is in our server.
+    in_guild = Column(Boolean, nullable=False, default=True)
+
+    # IDs of roles the user has on the server
+    roles = Column(ARRAY(BigInteger()), nullable=False)
+
+    @validates('id')
+    def validate_user_id(self, _key: str, user_id: int) -> Union[int, NoReturn]:
+        """Raise ValueError if the provided id is negative."""
+        if user_id < 0:
+            raise ValueError("User IDs cannot be negative.")
+        return user_id
+
+    @validates('discriminator')
+    def validate_discriminator(self, _key: str, discriminator: int) -> Union[int, NoReturn]:
+        """Raise ValueError if the provided discriminator is exceeds `9999`."""
+        if discriminator > 9999:
+            raise ValueError("Discriminators may not exceed `9999`.")
```

Comment:
```suggestion
        if discriminator > 9999 or discriminator < 0:
            raise ValueError("Discriminators may not exceed `9999` or be below `0000`.")
```

Reply:
`0` is an invalid discriminator too though, isn't it?

Reply:
It is, It's more like a check to make sure you don't pass a negative value.
Not so necessary, but a good thing to have in my opinion.

Reply:
Yes, but if we already do the check, then I think we should also do it correctly.

Reply:
On second thought, I don't think that, it is.  #0000 is preserved for webhooks, however I'm not sure why I left it in just because of webhooks. Will remove it.